---
layout: post
title: ECMAScript6 - <strong>Promise</strong>
subtitle: 프로미스
categories: es6
section: es6
description: Promise는 비동기 처리가 성공(fulfilled)하였는지 또는 실패(rejected)하였는지 등의 상태 정보와 처리 종료 후 실행될 콜백함수(then, catch)담고 있는 객체이다.
---

* TOC
{:toc}

![promise](./img/promise.jpg)

# 1. Promise와 콜백 헬(Callback Hell)

<!--처리가 종료하면 호출될 함수(콜백함수)를 미리 인자로 전달하고 처리가 종료하면 콜백함수를 호출하는 [비동기식 처리 모델](./js-async)은 JavaScript에서 빈번히 사용된다.

```javascript
asyncFunc(param, function(result) {
  /*
  처리가 종료하면 호출될 콜백함수
  result에 처리 결과가 전달된다
  */
});
```-->

JavaScript에서 빈번히 사용되는 [비동기 처리 모델](./js-async)은 요청을 병렬로 처리하여 다른 요청이 blocking(작업 중단)되는 않는 장점이 있지만 단점도 가지고 있는데 그것은 여러개의 콜백함수가 순서를 보장하기 위해 nesting되어 복잡도가 높아지는 <strong>Callback Hell</strong>이다.

![callback hell](./img/callback-hell.png)

Callback Hell
{: .desc-img}

<!-- ```javascript
step1(function(value1) {
  step2(value1, function(value2) {
    step3(value2, function(value3) {
      step4(value3, function(value4) {
        step5(value4, function(value5) {
            // Do something with value5
        });
      });
    });
  });
});
``` -->

Callback Hell은 코드의 가독성을 나쁘게 하고 복잡도를 증가시켜 실수를 유발시킬 확률이 높아지며 <strong>에러 처리가 곤란</strong>하다.

에러 처리가 곤란한 것은 콜백 방식 비동기 처리가 갖는 문제이다. exception은 caller 방향으로 전파된다. 그리고 비동기 처리의 콜백함수는 이벤트 큐에 들어가 있다가 해당 이벤트가 발생하면 호출된다.(이것에 대한 상세한 내용은 [이벤트 루프와 동시성(Concurrency)](http://localhost:4000/js-async#4-이벤트-루프와-동시성concurrency)을 참고하기 바란다) 이때 콜백함수를 호출한 것은 콜백함수를 갖는 비동기 함수가 아니기 때문에 아래와 같은 에러는 catch되지 않아 프로세스가 종료된다.

```javascript
try {
  setTimeout(function(){
    throw 'Error!';
  }, 1000);
} catch(e) {
  console.log('에러를 캐치하지 못한다..');
  console.log(e);
}
```

이러한 문제를 극복하기 위해 [Promise](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Promise)가 제안되었다. <!--(ES6의 [Generator](http://wiki.ecmascript.org/doku.php?id=harmony:generators)를 사용하여 Callback Hell 문제를 없애는 방법도 있다.)--> Promise는 ES6에 정식 채택되어 2017년 1월 현재 IE를 제외한 대부분의 브라우저가 지원하고 있다.

# 3. Promise의 상태(State)

Promise는 비동기 처리가 성공(fulfilled)하였는지 또는 실패(rejected)하였는지 등의 상태(state) 정보를 갖는다.

| 상태       | 의미                            | 구현
|-----------|--------------------------------|--------------
| pending   | 비동기 처리가 아직 수행되지 않은 상태   | resolve 또는 reject 함수가 아직 호출되지 않은 상태
| fulfilled | 비동기 처리가 성공적으로 수행된 상태    | resolve 함수가 호출된 상태 
| rejected  | 비동기 처리가 실패한 상태            | reject 함수가 호출된 상태 
| settled   | 비동기 처리가 수행된 상태(성공 또는 실패) | resolve 또는 reject 함수가 호출된 상태

# 2. Promise의 생성

Promise는 Promise 생성자를 통해 인스턴스화한다. Promise 생성자는 비동기 작업을 수행할 콜백함수를 인자로 전달받는데 이 콜백함수는 resolve와 reject 콜백함수를 인수로 전달받는다. 

```javascript
var promise = new Promise(function(resolve, reject) {
  // 비동기 작업 수행

  if (/* 비동기 작업 수행 성공 */) {
    resolve('resolved!');
  }
  else { /* 비동기 작업 수행 실패 */
    reject(Error('rejected!'));
  }
});
```

Promise 생성자가 인자로 전달받은 콜백 함수는 비동기 작업을 수행한다. 이때 비동기 작업이 성공하면 resolve를 호출하고 실패하면 reject를 호출한다.

# 4. Promise 후속 처리 함수 then, catch

Promise 생성자가 인자로 전달받은 콜백 함수에서 비동기 작업(Timer 함수)을 실행하도록 보자.

```javascript
// Promise 선언
function asyncFunc(param) {
  return new Promise(function(resolve, reject) {
    setTimeout(function() { // 비동기 함수
      param ? resolve("resolved!") : reject("rejected!");
    }, 2000);
  });
};
```

asyncFunc 함수는 함수 내부에서 Promise를 생성하고 반환하도록 선언되었다. 이제 asyncFunc 함수를 실행하여보자. asyncFunc 함수는 Promise를 반환하는데 이 Promise는 상태를 갖는다고 하였다. Promise의 상태에 따라 후속 처리 함수(then, catch)를 체이닝 방식으로 호출한다.

then
: then 메서드는 두 개의 콜백 함수를 인자로 전달 받는다. 첫번째 함수는 성공(fulfilled) 시 호출되는 함수이고 두번째 함수는 실패(rejected) 시 호출된다.

catch
: 예외 발생 시 호출된다.

```javascript
// Promise 선언
function asyncFunc(param) {
  return new Promise(function(resolve, reject) {
    setTimeout(function() { // 비동기 함수
      param ? resolve("resolved!") : reject("rejected!");
    }, 2000);
  });
};

// Promise 실행
asyncFunc(true)
  .then(function(data) {
    // resolve가 실행된 경우(성공), resolve 함수에 전달된 값이 data에 저장된다
    console.log(data);
  },function(reason) {
    // reject가 실행된 경우(실패), reject 함수에 전달된 값이 reason에 저장된다
    console.log(reason);
    throw "Error:" + reason;
  }).catch(function(error) {
    // 예외 발생 시 호출된다.
    console.error(error);
  });
```

위 예제는 비동기적 상황을 만들기 위해 Timer 함수를 사용하였지만 Promise는 XMLHttpRequest를 순서대로 처리하거나 처리 직후 다른 처리를 해야 할 때 유용하게 사용된다.

```javascript
var promise = new Promise(function(resolve, reject) {
  var xhr = new XMLHttpRequest();

  xhr.open('GET', 'http://poiemaweb.com/assets/data/data.html');

  xhr.onload = function() {
    if(this.status >= 200 && this.status < 300) { // 서버의 응답이 정상이면
      resolve(this.responseText);
    } else {
      reject(this.statusText);
    }
  };

  xhr.send(null);
});

promise.then(function(responseText) {
  console.log(responseText);
}).catch(function(statusText) {
  console.log('Error: ' + statusText);
});
```


# Reference

* [MDN: Promise](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Promise)

* [Google developers: Promise](https://developers.google.com/web/fundamentals/getting-started/primers/promises?hl=ko)
